Media player rendering with GL textures

ceyusa opened the pull request:
These patches pass the application's OpenGL raw context  and the its native display address to the media player, in order to create an internal wrapped context, thus it will generate video frames as textures.

For now only EGL from glutin-based app and android are in place, though tested only in Linux glutin app.

This PR also renders the generated frame textures by Servo/Media and renders them by using a thread that connects Webrenderer's ExternalImageHandler and each instantiated player. **By now, these patches, disable the WebGL rendering**. We need to provide a ExternalImageHandler demuxer.

This PR depends on https://github.com/servo/media/pull/270

- [X]  `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- This PR fixes #22300 and fixes #22920

In order to test it you must launch servo as

`./mach run -- --pref media.glvideo.enabled [...]`

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/23483)
<!-- Reviewable:end -->

ceyusa committed:
Fix code style

ceyusa committed:
Implement Window's GL context for Glutin

ferjm committed:
Implement WindowMethods for Android

highfive commented:
Heads up! This PR modifies the following files:
 * @asajeffrey: components/constellation/pipeline.rs, components/constellation/constellation.rs, components/script/dom/htmlmediaelement.rs, components/script/script_thread.rs, components/script/dom/window.rs
 * @cbrewster: components/constellation/pipeline.rs, components/constellation/constellation.rs
 * @paulrouget: components/constellation/pipeline.rs, ports/glutin/context.rs, components/compositing/Cargo.toml, ports/glutin/headless_window.rs, components/constellation/constellation.rs and 4 more
 * @kichjang: components/script/dom/htmlmediaelement.rs, components/script/script_thread.rs, components/script/dom/window.rs, components/script_traits/Cargo.toml, components/script_traits/lib.rs

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify **unsafe code**. Please review it carefully!
* These commits modify script code, but no tests are modified. Please consider adding a test!

bors-servo commented:
:umbrella: The latest upstream changes (presumably #23485) made this pull request unmergeable. Please resolve the merge conflicts.

ceyusa committed:
Add media.glvideo preference

bors-servo commented:
:umbrella: The latest upstream changes (presumably #23506) made this pull request unmergeable. Please resolve the merge conflicts.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #23517) made this pull request unmergeable. Please resolve the merge conflicts.

jdm commented:
Is this waiting for review right now?

ceyusa commented:
> Is this waiting for review right now?

Is still a working progress :( 

It would be great a glance to check if the sketched ideas make senss.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #23154) made this pull request unmergeable. Please resolve the merge conflicts.

ceyusa committed:
Add GLPlayerThreads in canvas

This factory will launch a GLPlayerThread from the application main
thread.

And add GLPlayerThread, the multiplexor for media players' video
renderers. This thread will receive commands from htmlmedialement
and webrenderer.

This code is also inspired by webgl_threads and WebGLThread.

ceyusa committed:
Pass GLPlayerThreads to constellation

Create the thread only if the GL context is known.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #23468) made this pull request unmergeable. Please resolve the merge conflicts.

ceyusa committed:
Define the channel type for player events

ceyusa committed:
Add GLPlayerMsgForward enum

These are the messages which are going to be sended/forwarded to
the player.

ceyusa committed:
GLPlayer thread API implementation

ceyusa committed:
Register and unregister a player in GLPlayer thread

ceyusa committed:
Add texture image in MediaFrameRenderer

ceyusa committed:
Implement webrender::ExternalImageHandler for player

Added trait GLPlayerExternalImageApi and its implementation
Implemented webrender::ExternalImageHandler using
GLPlayerExternalImageApi

ceyusa committed:
Instanciate and use the implemented webrender::ExternalImageHandler

ceyusa committed:
Add a task source for messages from GLPlayer thread

ceyusa committed:
Handle WR's lock/unlock logic under FrameRenderer

ceyusa committed:
Ignore frame if it is a GL texture

And mock what would be if we handle GL textures.

ceyusa committed:
Add GLPlayerMsg and glplayer_channel in canvas_traits

GLPlayerMsg enum values are going to be the commands to send to
the glplayer_thread.

glplayer_channel mod is a copy of webgl_channel.

ceyusa commented:
Ready for review :)

ceyusa committed:
Add more raw context handler for glutin port

ceyusa committed:
Add media (WindowGLContext) module in canvas_trait

This module adds a structure (WindowGLContext) which holds the
OpenGL parameters that are going to be used by servo-media player
to render video frames using OpenGL.

In order to fill this structure, three new methods were added to
WindowMethods trait. In this patch only the Glutin-based
implementation provides a simple boilerplate.

The WindowGLContext is created in the entry point of libservo, when
the application window is created, and later passed to the
constellation, the pipeline and to the window element in dom, thus
htmlmediaelement has a mean to obtain these parameters via its
window.

ceyusa committed:
Create the WindowGLContext at create_constellation()

ferjm committed:
Fix OSX build

bors-servo commented:
:umbrella: The latest upstream changes (presumably #23600) made this pull request unmergeable. Please resolve the merge conflicts.

ferjm committed:
Media crate

jdm commented on components/script/dom/htmlmediaelement.rs:
```diff
@@ -83,11 +83,12 @@ use std::mem;
 use webrender_api::{RenderApiSender, Transaction};
 
 pub struct MediaFrameRenderer {
+    id: u64,
```
Let's use an Option and a more descriptive name. player_id?

jdm commented on components/script/dom/htmlmediaelement.rs:
```diff
@@ -156,17 +158,17 @@ impl FrameRenderer for MediaFrameRenderer {
-                    //     None,
-                    // );
+                } else if self.id != 0 {
+                    txn.add_image(
```
Let's merge these blocks:
```rust
let image_data = if let Some(id) = self.player_id {
   ImageData::External(...)
} else {
   ImageData::Raw(frame.get_data())
};
txn.add_image(new_image_key, descriptor, image_data, None);
```

jdm commented on components/script/dom/htmlmediaelement.rs:
```diff
@@ -184,17 +186,17 @@ impl FrameRenderer for MediaFrameRenderer {
-                    //     }),
-                    //     None,
-                    // );
+                } else if self.id != 0 {
```
Some comments as previous case - let's merge these blocks.

jdm reviewed (changes requested):
Exciting! This looks like a sensible implementation.

jdm commented on ports/glutin/headless_window.rs:
```diff
@@ -195,4 +196,16 @@ impl WindowMethods for Window {
 
     fn prepare_for_composite(&self) { }
+
+    fn get_gl_context(&self) -> MediaPlayerCtxt::GlContext {
```
The point of this comment is to spur conversation - it feels like something we _should_ test if possible; how much work is it to support OSMesa for the GL video rendering integration?

ceyusa replied:
I have never tried GstGL on OSMesa. So far I have no idea.

ferjm committed:
Webrender external image handler demux

jdm commented:
Are the webrender external image changes ready for review? I saw some comments that suggested that some more work is expected.

ferjm commented:
Not ready for review yet. I need to add a few more changes.

El mar., 25 jun. 2019 21:01, Josh Matthews <notifications@github.com>
escribió:

> Are the webrender external image changes ready for review? I saw some
> comments that suggested that some more work is expected.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/servo/servo/pull/23483?email_source=notifications&email_token=AADVHSVS6IHP7R5IL5UHEKDP4JTQ3A5CNFSM4HRFVU42YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODYRIICI#issuecomment-505578505>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AADVHSTUAH7EG2DR6V26OWDP4JTQ3ANCNFSM4HRFVU4Q>
> .
>

bors-servo commented:
:umbrella: The latest upstream changes (presumably #23636) made this pull request unmergeable. Please resolve the merge conflicts.

ferjm committed:
Introduce WebrenderExternalImageRegistry

ferjm commented on components/script/dom/window.rs:
```diff
@@ -300,6 +301,10 @@ pub struct Window {
     layout_is_busy: Arc<AtomicBool>,
+
+    /// Window's GL context from application
+    #[ignore_malloc_size_of = "defined in script_thread"]
```
`WindowGLContext` has `GlContext`, `GlApi` and `NativeDisplay` members that are defined in servo-media. The 4th member is a `GLPlayerPipeline`, which is the only one defined in Servo, but it is basically an `IpcSender` or a `mpsc::Sender`, also defined outside of Servo.

ferjm committed:
Fix typos, warnings and other nits

ferjm commented:
@bors-servo try=wpt

bors-servo commented:
:hourglass: Trying commit 1b25a9ec10ee5c327a9d8283d32ec6a07002f75c with merge d1636eebc2fc145b01f88a9397bdc31c4fefebb5...

ferjm commented:
r? @jdm

bors-servo commented:
:broken_heart: Test failed - [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/12672)

ceyusa commented on components/webrender_traits/lib.rs:
```diff
@@ -0,0 +1,108 @@
+                },
+            };
+            webrender::ExternalImage {
+                uv: webrender_api::TexelRect::new(0.0, 0.0, size.width as f32, size.height as f32),
```
Sadly, textures from media and textures from WebGL don't have the same coordinates in texel space:

webgl: 0.0, size.height as f32, size.width as f32, 0.0

media (as gstreamer): 0.0, 0.0, size.width as f32, size.height as f32

bors-servo commented:
:umbrella: The latest upstream changes (presumably #23575) made this pull request unmergeable. Please resolve the merge conflicts.

ferjm committed:
Do not use WR external images if frames are not textures

ferjm commented:
@bors-servo try=wpt

bors-servo commented:
:hourglass: Trying commit 7b597ca07ab7abb0ac1d32d193bb096808641266 with merge 020433d821ba249b67fbfc88a039e80825f54cd6...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/12747)

ferjm committed:
Differentiate texel space coordinates for webgl and media

ferjm commented:
@bors-servo try=wpt

bors-servo commented:
:hourglass: Trying commit 216e1e500d191df7baf124c2015651d6a1f28ac3 with merge 86a8a392767a575d06efb37e912cfea3095cf296...

bors-servo commented:
:sunny: Test successful - [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/12687), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/12749), [status-taskcluster](https://tools.taskcluster.net/task-group-inspector/#/CNhiyeyKQ5u0lzwsS7hCPg)
State: approved= try=True

jdm commented on components/media/lib.rs:
```diff
@@ -2,12 +2,99 @@
-    GLPlayerChan, GLPlayerMsg, GLPlayerPipeline, GLPlayerReceiver, GLPlayerSender,
-};
+#![crate_name = "media"]
+#![crate_type = "rlib"]
```
These crate attributes shouldn't be necessary.

jdm commented on components/webrender_traits/lib.rs:
```diff
@@ -0,0 +1,108 @@
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+#![crate_name = "webrender_traits"]
+#![crate_type = "rlib"]
```
These crate attributes shouldn't be necessary.

jdm commented on components/webrender_traits/lib.rs:
```diff
@@ -0,0 +1,108 @@
+    /// Unlock the external image. The WR should not read the image
+    /// content after this call.
+    fn unlock(&mut self, key: webrender_api::ExternalImageId, _channel_index: u8) {
+        if let Some(handler_type) = self.external_images.get(&key) {
```
Let's use `expect` here instead.

jdm commented on components/webrender_traits/lib.rs:
```diff
@@ -0,0 +1,108 @@
+        _channel_index: u8,
+        _rendering: webrender_api::ImageRendering,
+    ) -> webrender::ExternalImage {
+        if let Some(handler_type) = self.external_images.get(&key) {
```
Let's use `expect` here instead.

jdm reviewed (approved):
Looks good!

bors-servo commented:
:umbrella: The latest upstream changes (presumably #23690) made this pull request unmergeable. Please resolve the merge conflicts.

ferjm committed:
Final tweaks: use expect and remove unnecessary crate attributes

ferjm commented:
@bors-servo r=jdm

bors-servo commented:
:pushpin: Commit 208473c has been approved by `jdm`

<!-- @bors-servo r=jdm 208473cdbce1e125338b461a3f0d55fb9d0bdd7c -->

bors-servo commented:
:hourglass: Testing commit 208473cdbce1e125338b461a3f0d55fb9d0bdd7c with merge 4a5c30257315cafc4bfe6ac71ad88f79197567a2...

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://tools.taskcluster.net/task-group-inspector/#/YhJNs8LjRQa5xmo1bIKhrA)

jdm commented:
Building simpleservo:
```
[0m[1m[38;5;9merror[E0063][0m[0m[1m: missing fields `gl_context_pointer`, `native_display_pointer` in initializer of `simpleservo::InitOptions`[0m
[0m   [0m[0m[1m[38;5;12m--> [0m[0mports/libsimpleservo/capi/src/lib.rs:120:16[0m
[0m    [0m[0m[1m[38;5;12m|[0m
[0m[1m[38;5;12m120[0m[0m [0m[0m[1m[38;5;12m| [0m[0m    let opts = InitOptions {[0m
[0m    [0m[0m[1m[38;5;12m| [0m[0m               [0m[0m[1m[38;5;9m^^^^^^^^^^^[0m[0m [0m[0m[1m[38;5;9mmissing `gl_context_pointer`, `native_display_pointer`[0m

[0m[1m[38;5;9merror[0m[0m[1m: aborting due to previous error[0m
```

Building windows:
```
error: unused import: `servo::gl::GlesFns`
  --> ports\libsimpleservo\api\src\gl_glue.rs:12:9
   |
12 |     use servo::gl::GlesFns;
   |         ^^^^^^^^^^^^^^^^^^
   |
   = note: `-D unused-imports` implied by `-D warnings`

error: unused import: `std::ffi::CString`
  --> ports\libsimpleservo\api\src\gl_glue.rs:13:9
   |
13 |     use std::ffi::CString;
   |         ^^^^^^^^^^^^^^^^^

error: unused import: `std::os::raw::c_void`
  --> ports\libsimpleservo\api\src\gl_glue.rs:14:9
   |
14 |     use std::os::raw::c_void;
   |         ^^^^^^^^^^^^^^^^^^^^

error: aborting due to 3 previous errors

error: Could not compile `simpleservo`.
```

Building magic leap:
```
error[E0599]: no method named `clone` found for type `simpleservo::gl_glue::egl::EGLInitResult` in the current scope
   --> ports/libmlservo/src/lib.rs:148:52
    |
148 |             MagicLeapVRService::new(name, ctxt, gl.clone()).expect("Failed to create VR service");
    |                                                    ^^^^^

error[E0063]: missing fields `gl_context_pointer`, `native_display_pointer` in initializer of `simpleservo::InitOptions`
   --> ports/libmlservo/src/lib.rs:153:16
    |
153 |     let opts = InitOptions {
    |                ^^^^^^^^^^^ missing `gl_context_pointer`, `native_display_pointer`

error[E0308]: mismatched types
   --> ports/libmlservo/src/lib.rs:175:29
    |
175 |     simpleservo::init(opts, gl, wakeup, callbacks).expect("error initializing Servo");
    |                             ^^ expected struct `std::rc::Rc`, found struct `simpleservo::gl_glue::egl::EGLInitResult`
    |
    = note: expected type `std::rc::Rc<(dyn gleam::gl::Gl + 'static)>`
               found type `simpleservo::gl_glue::egl::EGLInitResult`

error: aborting due to 3 previous errors

Some errors have detailed explanations: E0063, E0308, E0599.
For more information about an error, try `rustc --explain E0063`.
error: Could not compile `libmlservo`.
```

ferjm committed:
Fix libsimpleservo build

ferjm committed:
Fix Windows build

ferjm commented:
@bors-servo try=windows

bors-servo commented:
:hourglass: Trying commit f09661b36e7afa6ddaf8b3f4ecca3dd6ef4203e0 with merge 802efcdec3fb4d766a90beca4ecb74174f3295e6...

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://tools.taskcluster.net/task-group-inspector/#/at2bGcl1Tkag_CFRlmi2Eg)

jdm commented:
You can test the simpleservo changes locally using `./mach build -d --libsimpleservo`.

ferjm commented:
Yes, that worked on my laptop (OSX)

```
~/dev/mozilla/servo player-context*
❯ ./mach build -d --libsimpleservo
   Compiling simpleservo v0.0.1 (/Users/ferjm/dev/mozilla/servo/ports/libsimpleservo/api)
   Compiling simpleservo_capi v0.0.1 (/Users/ferjm/dev/mozilla/servo/ports/libsimpleservo/capi)
    Finished dev [unoptimized + debuginfo] target(s) in 2m 41s
Build Completed in 0:02:41
```

jdm commented:
@bors-servo try=windows

bors-servo commented:
:hourglass: Trying commit 23e4812b41ffb5c1154b5dd4f7c5bb2285551198 with merge 784945798ea324d998ee7bcf3ebd89637d31da55...

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://tools.taskcluster.net/task-group-inspector/#/cAd7h_5USmGUKy1hN4-Fgg)

jdm committed:
Fix magic leap build.

jdm commented:
@bors-servo r+

bors-servo commented:
:pushpin: Commit 5d30716 has been approved by `jdm`

<!-- @bors-servo r=jdm 5d307160b4be4abb4b986cdc63d6650ae2b32895 -->

bors-servo commented:
:hourglass: Testing commit 5d307160b4be4abb4b986cdc63d6650ae2b32895 with merge 320ec43f1b6794ac8c818db00412da8819b29117...

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://tools.taskcluster.net/task-group-inspector/#/V6VPJ46DSP2GKGeyX1yt-A)

CYBAI commented:
In `magicleap` build:

```
error[E0560]: struct `simpleservo::InitOptions` has no field named `gl_context_points`
   --> ports/libmlservo/src/lib.rs:160:9
    |
160 |         gl_context_points: Some(gl.gl_context),
    |         ^^^^^^^^^^^^^^^^^ help: a field with a similar name exists: `gl_context_pointer`

error[E0560]: struct `simpleservo::InitOptions` has no field named `display`
   --> ports/libmlservo/src/lib.rs:161:9
    |
161 |         display: Some(gl.display),
    |         ^^^^^^^ `simpleservo::InitOptions` does not have this field
    |
    = note: available fields are: `args`, `url`, `coordinates`, `density`, `vr_init` ... and 3 others

error: aborting due to 2 previous errors
```

jdm commented:
@bors-servo r+

bors-servo commented:
:pushpin: Commit 4fe1291 has been approved by `jdm`

<!-- @bors-servo r=jdm 4fe129109bbc40a32e491825d77a67fa8e5e9662 -->

bors-servo commented:
:hourglass: Testing commit 4fe129109bbc40a32e491825d77a67fa8e5e9662 with merge 0dc17af7f0dab01f673894b6ac7c84cb50d03603...

bors-servo commented:
:sunny: Test successful - [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/12728), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/12790), [status-taskcluster](https://tools.taskcluster.net/task-group-inspector/#/EOQQN2vgThG9RNDEBg74Wg)
Approved by: jdm
Pushing 0dc17af7f0dab01f673894b6ac7c84cb50d03603 to master...
